modules/imgcodecs/src/grfmt_tiff.cpp

```diff
@@ -804,6 +804,22 @@ bool  TiffEncoder::writeLibTiff( const Mat& img, const std::vector<int>& params)
         return false;
     }
 
+    //Get dpi.
+    int dpix = 300;
+    int dpiy = 300;
+    int resUnit = RESUNIT_INCH;
+    readParam(params, TIFFTAG_XRESOLUTION, dpix);
+    readParam(params, TIFFTAG_YRESOLUTION, dpiy);
+    readParam(params, TIFFTAG_RESOLUTIONUNIT, resUnit);
```

Comment:
"params" are imwrite's values from this enum: https://github.com/opencv/opencv/blob/3.4.0/modules/imgcodecs/include/opencv2/imgcodecs.hpp#L81

Using values outside of mentioned enum just violates the current OpenCV `imwrite()` design (conflicts in enum values are possible so these hacks should not be allowed at all).

Also OpenCV public API should not rely on external "custom" headers, like "tiff.h".

Problem is here and for other "readParam()" calls. Related #361